skbio/metadata/_interval.py

```diff
@@ -0,0 +1,334 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2013--, scikit-bio development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from ._intersection import IntervalTree
+
+
+class Interval():
+    '''Stores the position and metadata of an interval.
+
+    Parameters
+    ----------
+    intervals : list of tuple of ints
+        List of tuples representing start and end coordinates.
+    boundaries : list of tuple of bool
+        List of tuples, representing the openness of each interval.
+    metadata : dict
+        Dictionary of attributes storing information of the feature
+        such as `strand`, `gene_name` or `product`.
+    interval_metadata : object
+        A reference to the `IntervalMetadata` object that this
+        interval is associated to.
+    '''
+    def __init__(self, intervals=None, boundaries=None,
+                 metadata=None, interval_metadata=None):
+        iv = []
+        for interval in intervals:
+            iv.append(_polish_interval(interval))
+
+        if boundaries is not None:
+            self.boundaries = boundaries
+        else:
+            self.boundaries = []
+
+        if metadata is not None:
+            self.metadata = metadata
+        else:
+            self.metadata = {}
+
+        self._interval_metadata = interval_metadata
+
+        if intervals is not None:
+            self.intervals = sorted(iv)
+        else:
+            self.intervals = []
+
+        if interval_metadata is not None:
+            self._add_interval()
+
+    def _add_interval(self):
+        """ Add an interval object to the IntervalTree within
+            IntervalMetadata."""
+        # Add directly to the tree.  So no need for _is_stale_tree
+        for loc in self.intervals:
+            if loc is not None:
+                start, end = loc
+                self._interval_metadata._intervals.add(start, end, self)
+        self._interval_metadata._metadata.append(self)
+
+    def __getitem__(self, key):
+        return self.metadata[key]
+
+    def __setitem__(self, key, val):
+        self.metadata[key] = val
+
+    # This is required for creating unique sets of intervals
+    def __hash__(self):
```

Comment:
Here is an example to illustrate this

```python
In [1]: from skbio.metadata import IntervalMetadata

In [2]: %paste
        im = IntervalMetadata()
        im.add(intervals=[(0, 2), (4, 7)],
               metadata={'gene': 'sagA', 'location': 0})
        im.add(intervals=[(3, 5)],
               metadata={'gene': 'sagB', 'location': 0})

## -- End pasted text --

In [3]: im.query(intervals=[(1,5)])
Out[3]: 
[ Interval(intervals=[(0, 2), (4, 7)], metadata={'location': 0, 'gene': 'sagA'}),
  Interval(intervals=[(3, 5)], metadata={'location': 0, 'gene': 'sagB'}),
  Interval(intervals=[(0, 2), (4, 7)], metadata={'location': 0, 'gene': 'sagA'})]
```

One of the features gets outputted twice, because it is over a non contiguous set of intervals.  Some sort of uniqueness operation should be applied.  A hash isn't necessary - but it is the a very easy route.  Now a sort would do just fine - but we'd need all of the comparison operators to be in place.